### PR TITLE
Add sign verification for unzip

### PR DESCRIPTION
**Pull request by zhili1208**
let's review those existing code.... 

For packages.config, there are still some works which need to be done[clean up when package is invalid]

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+using System;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs]: Add .NET license header.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        public XmlDocFileSaveMode XmlDocFileSaveMode { get; set; }
+
+        public SignedPackageVerifier SignedPackageVerifier { get; }
+
+        public PackageExtractionContextBase(
+            PackageSaveMode packageSaveMode,
+            XmlDocFileSaveMode xmlDocFileSaveMode,
+            ILogger logger,
+            SignedPackageVerifier signedPackageVerifier
+            )
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs]: Move to end of previous line.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            Logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            PackageSaveMode = packageSaveMode;
+            XmlDocFileSaveMode = xmlDocFileSaveMode;
+            SignedPackageVerifier = signedPackageVerifier;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs]: Should you throw an `ArgumentNullException` if `signedPackageVerifier` is `null`?

File: src/NuGet.Clients/NuGet.CommandLine/Common/ConsoleProjectContext.cs [hunk 4]
```diff
@@ -1,4 +1,4 @@
-﻿using System.Globalization;
+using System.Globalization;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.CommandLine/Common/ConsoleProjectContext.cs]: Add .NET license header.

File: src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs [hunk 5]
```diff
@@ -301,7 +302,11 @@ private NuGetPackageManager GetNuGetPackageManager(string solutionDirectory)
             // and not have to determine if the package is a satellite package beforehand
             if (nuGetProjectContext.PackageExtractionContext == null)
             {
-                nuGetProjectContext.PackageExtractionContext = new PackageExtractionContext(new LoggerAdapter(nuGetProjectContext));
+                var signedPackageVerifier = new SignedPackageVerifier(
+                           SignatureVerificationProviderFactory.GetSignatureVerificationProviders(),
+                           SignedPackageVerifierSettings.Default);
+
+                nuGetProjectContext.PackageExtractionContext = new PackageExtractionV2Context(new LoggerAdapter(nuGetProjectContext), signedPackageVerifier: signedPackageVerifier);
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.PackageManagement/IDE/PackageRestoreManager.cs]: Remove the named parameter `signedPackageVerifier`; it's unnecessary, and the local variable is already well-named.

File: src/NuGet.Core/NuGet.PackageManagement/Projects/FolderNuGetProject.cs [hunk 6]
```diff
@@ -145,7 +146,11 @@ public override Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(Ca
                     var packageExtractionContext = nuGetProjectContext.PackageExtractionContext;
                     if (packageExtractionContext == null)
                     {
-                        packageExtractionContext = new PackageExtractionContext(new LoggerAdapter(nuGetProjectContext));
+                        var signedPackageVerifier = new SignedPackageVerifier(
+                            SignatureVerificationProviderFactory.GetSignatureVerificationProviders(),
+                            SignedPackageVerifierSettings.Default);
+
+                        packageExtractionContext = new PackageExtractionV2Context(new LoggerAdapter(nuGetProjectContext), signedPackageVerifier: signedPackageVerifier);
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.PackageManagement/Projects/FolderNuGetProject.cs]: Remove `signedPackageVerifier` named parameter.

File: src/NuGet.Core/NuGet.PackageManagement/Projects/FolderNuGetProject.cs [hunk 7]
```diff
@@ -399,7 +404,11 @@ public bool PackageAndManifestExists(PackageIdentity packageIdentity)
             var packageExtractionContext = nuGetProjectContext.PackageExtractionContext;
             if (packageExtractionContext == null)
             {
-                packageExtractionContext = new PackageExtractionContext(new LoggerAdapter(nuGetProjectContext));
+                var signedPackageVerifier = new SignedPackageVerifier(
+                            SignatureVerificationProviderFactory.GetSignatureVerificationProviders(),
+                            SignedPackageVerifierSettings.Default);
+
+                packageExtractionContext = new PackageExtractionV2Context(new LoggerAdapter(nuGetProjectContext), signedPackageVerifier: signedPackageVerifier);
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.PackageManagement/Projects/FolderNuGetProject.cs]: Remove `signedPackageVerifier` named parameter.

File: src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs [hunk 8]
```diff
@@ -54,6 +55,19 @@ public static class PackageExtractor
                 var packageDirectoryInfo = Directory.CreateDirectory(installPath);
                 var packageDirectory = packageDirectoryInfo.FullName;
 
+                if (packageExtractionV2Context.SignedPackageVerifier != null)
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/PackageExtractor.cs]: This block of code is duplicated several times throughout this file.  How about extracting it into a helper method?

File: src/NuGet.Core/NuGet.Packaging/Signing/SignatureVerificationProviderFactory.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+using System.Collections.Generic;
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/SignatureVerificationProviderFactory.cs:4]: Add .NET license header.

File: src/NuGet.Core/NuGet.Protocol/Utility/OfflineFeedUtility.cs [hunk 10]
```diff
@@ -193,12 +194,17 @@ public static void ThrowIfInvalid(string path)
...
+                            SignedPackageVerifierSettings.Default);
+
+                        var packageExtractionV3Context = new PackageExtractionV3Context(
                             packageIdentity,
                             source,
                             logger,
                             packageSaveMode: packageSaveMode,
-                            xmlDocFileSaveMode: PackageExtractionBehavior.XmlDocFileSaveMode);
+                            xmlDocFileSaveMode: PackageExtractionBehavior.XmlDocFileSaveMode,
+                            signedPackageVerifier: signedPackageVerifier);
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Protocol/Utility/OfflineFeedUtility.cs]: Can simplify the above 3 lines to:

    packageSaveMode, 
    PackageExtractionBehavior.XmlDocFileSaveMode, 
    signedPackageVerifier);

File: test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs [hunk 11]
```diff
@@ -449,12 +449,13 @@ public class RestoreCommandTests
...
                         packageDownloader,
-                        new VersionFolderPathContext(
+                        new PackageExtractionV3Context(
                             new PackageIdentity("packageA", NuGetVersion.Parse("1.0.0")),
                             packagesDir,
                             logger,
                             PackageSaveMode.Defaultv3,
-                            XmlDocFileSaveMode.None),
+                            XmlDocFileSaveMode.None,
+                            null),
```
Comment (dtivel) [inline test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommandTests.cs]: Add parameter name for `null`.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs [hunk 12]
```diff
@@ -1173,5 +1174,75 @@ public class RestoreCommandTests
...
         }
+
+        [Fact]
+        public async Task RestoreCommand_InvalidSignPackage()
+        {
+            // Arrange
+            var sources = new List<PackageSource>();
+
+            // Both TxMs reference packageA, but they are different types.
+            // Verify that the reference does not show up under libraries.
```
Comment (dtivel) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs]: Does this comment make sense for this test?

File: test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageExtractorTests.cs [hunk 13]
```diff
@@ -206,13 +210,14 @@ public class PackageExtractorTests
...
-                        new VersionFolderPathContext(
+                        new PackageExtractionV3Context(
                             identity,
                             packagesPath,
                             isLowercase,
                             NullLogger.Instance,
                             PackageSaveMode.Nupkg,
-                            XmlDocFileSaveMode.None),
+                            XmlDocFileSaveMode.None,
+                            null),
```
Comment (dtivel) [inline test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageExtractorTests.cs]: Add parameter name explaining `null`.

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs [hunk 14]
```diff
@@ -352,9 +353,13 @@ private DependencyBehavior GetDependencyBehavior()
...
             else
             {
+                var signedPackageVerifier = new SignedPackageVerifier(
+                            SignatureVerificationProviderFactory.GetSignatureVerificationProviders(),
+                            SignedPackageVerifierSettings.Default);
+
                 var projectContext = new ConsoleProjectContext(Console)
                 {
-                    PackageExtractionContext = new PackageExtractionContext(Console)
+                    PackageExtractionContext = new PackageExtractionV2Context(Console, signedPackageVerifier)
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: Why add a new class if you're removing the PackageExtractionContext regardless?
Why not just change that class directly?

Comment (zhili1208) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: Good point, I should change the class directly

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: nvm...I just noticed that there's a PackageExtractionV3Context as well. 

I am not fan of this type of naming as it does not provide any additional meaning. 

Does V2/V3 have anything to do with the protocols themselves? 

It's just very confusing imo.

File: src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs (hunk 3 above)
Comment (zhili1208) [reply src/NuGet.Core/NuGet.Packaging/PackageExtraction/PackageExtractionContextBase.cs]: No, user can set signedPackageVerifier to null to skip signature verification

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs [hunk 15]
```diff
@@ -125,6 +125,11 @@ public enum NuGetLogCode
         NU1401 = 1401,
 
         /// <summary>
+        /// Package Signature is invalid
+        /// </summary>
+        NU1410 = 1410,
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:130]: This can be NU1402

**Pull request merged**